iscservice/demo/util.c

```diff
@@ -0,0 +1,710 @@
+}
+
+/* Warning: non-reentrant. */
+static char* fid2str(const struct m0_fid *fid)
+{
+	static char buf[256];
+
+	if (fid)
+		sprintf(buf, FID_F, FID_P(fid));
+	else
+		sprintf(buf, "%p", fid);
+
+	return buf;
+}
+
+/**
+ * Return parity group size for object.
+ */
+uint64_t m0util_m0gs(struct m0_obj *obj, int tier)
+{
+	int                     rc;
+	unsigned long           usz; /* unit size */
+	struct m0_reqh         *reqh = &m0_instance->m0c_reqh;
+	struct m0_pool_version *pver;
+	struct m0_fid          *pool = tier2pool(tier);
+
+	rc = m0_pool_version_get(reqh->rh_pools, pool, &pver);
+	if (rc != 0) {
+		ERR("m0_pool_version_get(pool=%s) failed: rc=%d\n",
+		    fid2str(pool), rc);
+		return 0;
+	}
+
+	usz = m0_obj_layout_id_to_unit_size(obj->ob_attr.oa_layout_id);
+
+	return usz * pver->pv_attr.pa_N;
+}
+
+void free_segs(struct m0_bufvec *data, struct m0_indexvec *ext,
+	       struct m0_bufvec *attr)
+{
+	m0_indexvec_free(ext);
+	m0_bufvec_free(data);
+	m0_bufvec_free(attr);
+}
+
+int alloc_segs(struct m0_bufvec *data, struct m0_indexvec *ext,
+	       struct m0_bufvec *attr, uint64_t bsz, uint32_t cnt)
+{
+	int i, rc;
+
+	rc = m0_bufvec_alloc(data, cnt, bsz) ?:
+	     m0_bufvec_alloc(attr, cnt, 1) ?:
+	     m0_indexvec_alloc(ext, cnt);
+	if (rc != 0)
+		goto err;
+
+	for (i = 0; i < cnt; i++)
+		attr->ov_vec.v_count[i] = 0; /* no attrs */
+
+	return 0;
+ err:
+	free_segs(data, ext, attr);
+	return rc;
+}
+
+uint64_t set_exts(struct m0_indexvec *ext, uint64_t off, uint64_t bsz)
+{
+	uint32_t i;
+
+	for (i = 0; i < ext->iv_vec.v_nr; i++) {
+		ext->iv_index[i] = off;
+		ext->iv_vec.v_count[i] = bsz;
+		off += bsz;
+	}
+
+	return i * bsz;
+}
+
+static int spiel_prepare(struct m0_spiel *spiel)
+{
+	struct m0_reqh *reqh = &m0_instance->m0c_reqh;
+	int             rc;
+
+	rc = m0_spiel_init(spiel, reqh);
+	if (rc != 0) {
+		fprintf(stderr, "error! spiel initialisation failed.\n");
+		return rc;
+	}
+
+	rc = m0_spiel_cmd_profile_set(spiel, m0_conf.mc_profile);
+	if (rc != 0) {
+		fprintf(stderr, "error! spiel initialisation failed.\n");
+		return rc;
+	}
+	rc = m0_spiel_rconfc_start(spiel, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "error! starting of rconfc failed in spiel failed.\n");
+		return rc;
+	}
+
+	return 0;
+}
+
+static void m0util_spiel_destroy(struct m0_spiel *spiel)
+{
+	m0_spiel_rconfc_stop(spiel);
+	m0_spiel_fini(spiel);
+}
+
+static bool conf_obj_is_svc(const struct m0_conf_obj *obj)
+{
+	return m0_conf_obj_type(obj) == &M0_CONF_SERVICE_TYPE;
+}
+
+struct m0_rpc_link * m0util_svc_rpc_link_get(struct m0_fid *svc_fid)
+{
+	struct m0_reqh *reqh = &m0_instance->m0c_reqh;
+	struct m0_reqh_service_ctx *ctx;
+	struct m0_pools_common *pc = reqh->rh_pools;
+
+	m0_tl_for(pools_common_svc_ctx, &pc->pc_svc_ctxs, ctx) {
+		if (m0_fid_eq(&ctx->sc_fid, svc_fid))
+			return &ctx->sc_rlink;
+	} m0_tl_endfor;
+
+	return NULL;
+}
+
+/*
+ * Loads a library into m0d instances.
+ */
+int m0util_isc_api_register(const char *libpath)
```

Comment:
Why do we need this?
Can we register the lib on servers, manually?
If not, can we register the lib on client manually?
if spiel must be used, we should add this API to Motr.

So, any isc app can just use this API. 
It should not be part of the demo. It should be part of Motr.

Reply:
Agree, let me try to put it into libmotr...

Reply:
Done.